src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCLoginBackingBean.java

```diff
@@ -0,0 +1,169 @@
+            setUser();
+            return SystemConfig.getDataverseSiteUrlStatic();
+        }
+        return "/oidc/login?target=JSF&oidcp=" + oidcAuthProvider.getId();
+    }
+
+    /**
+     * View action for callback.xhtml, the browser redirect target for the OAuth2
+     * provider.
+     * 
+     * @throws IOException
+     */
+    public void setUser() {
+        try {
+            final String subject = openIdContext.getSubject();
+            final OIDCAuthProvider provider = getProvider();
+            final UserRecordIdentifier userRecordIdentifier = new UserRecordIdentifier(provider.getId(), subject);
+            AuthenticatedUser dvUser = authenticationSvc.lookupUser(userRecordIdentifier);
+            if (dvUser == null) {
+                if (!systemConfig.isSignupDisabledForRemoteAuthProvider(provider.getId())) {
+                    final JwtClaims claims = openIdContext.getAccessToken().getJwtClaims();
+                    final String firstName = claims.getStringClaim(OpenIdConstant.GIVEN_NAME).orElse("");
+                    final String lastName = claims.getStringClaim(OpenIdConstant.FAMILY_NAME).orElse("");
+                    final String verifiedEmailAddress = getVerifiedEmail();
+                    final String emailAddress = verifiedEmailAddress == null ? "" : verifiedEmailAddress;
+                    final String affiliation = claims.getStringClaim("affiliation").orElse("");
+                    final String position = claims.getStringClaim("position").orElse("");
+                    final OAuth2UserRecord userRecord = new OAuth2UserRecord(
+                            provider.getId(),
+                            subject,
+                            claims.getStringClaim(OpenIdConstant.PREFERRED_USERNAME).orElse(subject),
+                            null,
+                            new AuthenticatedUserDisplayInfo(firstName, lastName, emailAddress, affiliation, position),
+                            List.of(emailAddress));
+                    logger.log(Level.INFO, "redirect to first login: " + userRecordIdentifier);
+                    newAccountPage.setNewUser(userRecord);
+                    Faces.redirect("/oauth2/firstLogin.xhtml");
+                }
+            } else {
+                dvUser = userService.updateLastLogin(dvUser);
+                session.setUser(dvUser);
+                Faces.redirect("/");
+            }
+        } catch (Exception e) {
+            logger.log(Level.SEVERE, "Setting user failed: " + e.getMessage());
+        }
+    }
+
+    public UserRecordIdentifier getUserRecordIdentifier() {
+        try {
+            final String subject = openIdContext.getSubject();
+            final OIDCAuthProvider provider = getProvider();
+            return new UserRecordIdentifier(provider.getId(), subject);
+        } catch (final Exception ignore) {
+            return null;
+        }
+    }
+
+    private String getVerifiedEmail() {
```

Comment:
The code in the method is not very readable. Can it be simplified?

Reply:
I tried...